ministryofjustice/correspondence_tool_staff
# CT-2960 CT-3005 edit and data requests table

## Description
This PR includes the work to display and edit data requests, as well as mark them complete. 

It also includes changes to the data table design that emerged as late requirements following user testing and feedback. 

You can:
* View the data request details entered on the New Data Request form using the new request_types and from/to dates when relevant
* Edit a data request to amend the location, request type, date requested, and from/to dates
* Edit a data request to update the number of pages received
* Edit a data request to update the status to completed 

Some things to observe
1. The status outputs as "Completed" or "In progress" depending on the completed status value
2. The from/to dates output in a row beneath the request type if they're present
3. If the from date is omitted, the line reads "Up to" the _to date_
4. If the to date is omitted, the line reads _from date_ "onwards"
5. When you edit the details of the data request, the Case History shows a generic status line
6. When you edit the pages received, the Case history shows a custom message with the old and new page counts
7. When you choose "Edit" on a data request and then save without any changes, the case show page reloads with a notice to say no changes were made and NO message appears in the case history

The data request log history is no longer updated when the page count changes. This is a known issue and there is no intent to fix it right now. When we come to the Chasing stories we will revisit the data request logs functionality. 

## Self-review checklist


### Screenshots

![image](https://user-images.githubusercontent.com/1161161/91048933-4e8fc200-e614-11ea-88ca-c072a4beda27.png)

![image](https://user-images.githubusercontent.com/1161161/91048959-5a7b8400-e614-11ea-916b-7452fd03ada9.png)

![image](https://user-images.githubusercontent.com/1161161/91049352-01602000-e615-11ea-8bb7-830f0581b6cc.png)

### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-3005
https://dsdmoj.atlassian.net/browse/CT-2960
https://dsdmoj.atlassian.net/browse/CT-2993

### Deployment
This PR includes database migrations that can be rolled back if necessary

`rails db:migrate`

### Manual testing instructions
Run the database migrations
Edit an existing Offender SAR
Create a new data request
Verify the details output appropriately
Verify the edit function allows you to edit the details
Verify the complete checkbox updates the status on the table
Verify the pages received mechanism works as intended
